### PR TITLE
(chore) Clean up all regexs to be UTF-8 compliant/ready

### PR DESCRIPTION
**Pull request by joshgoebel**
Work toward #2756.

Cleans up a lot of incorrect (unnecessary escaped) regex and would not compile with the `u` flag.  After that makes some rather large performance improvements (with utf8 turned on at least) to `yaml` and `mipsasm`.  It looks like the mipasm rules have been wrong all alone... as far as I can determine they are intended to match a literal `.` (otherwise they are far too broad) but were matching any character - which seems to terribly slow down the whole grammar in `u` mode.

This consisted mostly of:

- Most unescaped `{` and `}`
- Lots of unneeded escapes for `-`, `<`, `>`, and others.
- Converting strings to regex if it made them simpler, easier to read (editor syntax coloring)

I plan to review the PR myself line by line, but can't imagine it'll be fun.  I did try to change the minimal amount necessary.  Often I turned strings into regex if it made them easier to read and see what I was doing.  Once in a while I touched a nearby regex.

I imagine the fact that all tests still pass is a pretty good indication this is 99% correct. :-)

Note: This doesn't actually turn on UTF8 anywhere... it just fixes all the regex so that if `u` is added inside the main mode compiler everything "just works".  It is still needed to be reviewed what else might need to be done on the road to UTF8 support.

### CODE REVIEW

**Comment by joshgoebel** [comment]
No rush, I'm pushing this off until 10.4.  10.3 is big enough.

File: src/languages/qml.js
Comment (joshgoebel) [inline src/languages/qml.js:93]: Need to review this.

File: src/languages/tap.js
Comment (joshgoebel) [inline src/languages/tap.js:25]: This was broken before `\s` vs `\\s` and when fixed seemed to break things so without any context going to go with the simpler rule for now.

File: src/languages/dsconfig.js
Comment (joshgoebel) [inline src/languages/dsconfig.js:18]: I don't completely understand when `-` needs to be escaped inside a `[]` block since sometimes it seems to figure it out on context (or maybe it's my imagination).

File: src/languages/mipsasm.js
Comment (joshgoebel) [inline src/languages/mipsasm.js:93]: Need to look at this carefully.

File: src/languages/prolog.js
Comment (joshgoebel) [inline src/languages/prolog.js:60]: This is greek to me, but I think it's correct... anyone know Prolog?  The "\s" is literal?

File: src/languages/yaml.js
Comment (joshgoebel) [inline src/languages/yaml.js:105]: Speed optimization... I think this is accomplishing the same thing...

File: src/languages/routeros.js
Comment (joshgoebel) [inline src/languages/routeros.js:147]: Maybe I broke this?  Not sure what it's supposed to be doing.

File: src/languages/lsl.js
Comment (joshgoebel) [inline src/languages/lsl.js:34]: There are speed optimizations believe it or not the benchmark was faster without the ":" to mark it as a non capturing group.

File: src/languages/mipsasm.js
Comment (joshgoebel) [inline src/languages/mipsasm.js:93]:
```suggestion
    // forward slashes are no allowed
    illegal: /\//
```

Comment (joshgoebel) [reply src/languages/mipsasm.js:93]: Yep, it's the same as before.  Adding a comment to clarify.

File: src/lib/mode_compiler.js
Comment (joshgoebel) [inline src/lib/mode_compiler.js:27]: This will be reverted before merge.

File: src/highlight.js
Comment (joshgoebel) [inline src/highlight.js:582]: This will be reverted before merge.

**Comment by joshgoebel** [comment]
I've 99% review this myself but could still use another set of eyes. :)

File: src/languages/dsconfig.js
Comment (joshgoebel) [reply src/languages/dsconfig.js:18]: Ah, seems simple enough... in cases where it could be ambiguous it now wants it to be escaped:

```
regex = /[a\-b]/u
```

But if you just always put it at the end of the character group then it's obvious it's not a range.

File: src/languages/routeros.js
Comment (joshgoebel) [reply src/languages/routeros.js:147]: Pretty sure it's fine, added a comment to explain it.

File: src/languages/prolog.js
Comment (joshgoebel) [reply src/languages/prolog.js:60]: This makes sense to me now I think.  Code might be:

```prolog
% \s literal (escaped)
0'\s
% character "b"
0'b
% character "'" (escaped)
0'\'
```

File: src/languages/dsconfig.js
Comment (allejo) [reply src/languages/dsconfig.js:18]: From the changes I've seen up until this file, it looks like you've moved the `-` to the end of each `[]` or it was already there, do you want to keep that consistent?

File: src/languages/ebnf.js
Comment (allejo) [inline src/languages/ebnf.js:12]:
```js
begin: /^[ ]*[a-zA-Z][a-zA-Z-_]*([\s\-_]+[a-zA-Z][a-zA-Z]*)*/
#                           ^       ^^
```

You're escaping one `-` here but not the other?

File: src/languages/xquery.js
Comment (allejo) [inline src/languages/xquery.js:109]: hyphen consistency

File: src/languages/yaml.js
Comment (allejo) [reply src/languages/yaml.js:105]: `\S` is anything _but_ whitespace. So `[\S ]` to me sounds like "anything but whitespace but also allow a single space character". I don't think tabs or spaces aren't covered by `\n`?

File: src/languages/dsconfig.js
Comment (joshgoebel) [reply src/languages/dsconfig.js:18]: I'm happy now that I understand it.

File: src/languages/yaml.js
Comment (joshgoebel) [reply src/languages/yaml.js:105]: But we're using "not \n"... The question to ask here is what was the original intent... aren't we trying to handle code like:

```
key: |
  multi-string
  value
```

Ie, some number of spaces (matching, hence the `\2`) then content (up until a new line)...  I believe `[^\n]` matches the intention, yes?

Comment (joshgoebel) [reply src/languages/yaml.js:105]: The full expression is `[\S ]+`... not space or space, repeated at least once... and this naturally excludes new lines... so I (I believe) wrote an equivalent and faster version.

File: src/languages/ebnf.js
Comment (joshgoebel) [reply src/languages/ebnf.js:12]: Doesn't need to be, but I'll move them both to the end to make this consistent.

File: src/languages/gams.js
Comment (joshgoebel) [reply src/languages/gams.js:81]: - The `[` does not need to be escape inside a `[` block... it's clear it's not opening a new block because it's inside one.
- `<` is not a character that's has ever needed escaping (so says my linter), and it's now invalid in `/u` mode.

> Like technically /{/ doesn't need to be escaped, but this PR is escaping it anyways.

Very different `{` **must be escaped** for the stricter UTF-8, which is the whole point of this PR.  `[` does not need to be.  The overall preference here is not to use unnecessary escapes anymore.

File: src/languages/less.js
Comment (joshgoebel) [reply src/languages/less.js:127]: In this regard I don't think it matters, plus without a linter rule this is hard to enforce over time in any case.  I'm happy if it's escaped properly where-ever it's found.

File: src/languages/stylus.js
Comment (joshgoebel) [reply src/languages/stylus.js:125]: Thanks, hard to see all these. :)

File: src/languages/lisp.js
Comment (joshgoebel) [reply src/languages/lisp.js:11]: It's used with `+` and strings later, so it doesn't work if it's converted to a regex... I'd need to start using the `regex` tools everywhere... 

If someone wanted to focus on "regex not string" a lot more as a separate initiative I'd be game for that, but I only tried to do that here in cases where it made sense and was trivial to do - the scope of this PR is already huge.

File: src/languages/gams.js
Comment (allejo) [reply src/languages/gams.js:81]: > Very different `{` **must be escaped** for the stricter UTF-8, which is the whole point of this PR. `[` does not need to be. The overall preference here is not to use unnecessary escapes anymore. _They've always been flagged by the linter, it just hasn't been a high priority before now._

ah thanks for the explanation! Didn't realize that `{` had to be escaped for UTF-8 support. I just thought since it wasn't ambiguous it didn't need to be escaped.

---

> I'm a bit confused on this one. Is the removal of `\\` here correct? Is it trying to escape `<` or is `\` an actual character in this `[]` block?

Can confirm this? Being inside of a `[]` and regex, it looks to be like `\\` is for an `\` character and not trying to escape `<`. i.e. if this were a string originally, then I agree that it should be removed because it'd be `\<` but this has always been a regex so it doesn't look to me like it's escaping the `<`.

Comment (joshgoebel) [reply src/languages/gams.js]: Oh yikes, yeah I think you are right. :)  I put it back!

File: src/languages/prolog.js
Comment (allejo) [reply src/languages/prolog.js:60]: Yea, looks like `0'` is the way of getting the character code of a char. So `0'\s` returns `32`. And you are correct, `0''` is valid, it returns `39`

Screenshots and confirmation courtesy of @Adwitiya-Singh

image

More info, https://www.swi-prolog.org/pldoc/man?section=charescapes

Comment (joshgoebel) [reply src/languages/prolog.js:60]: > And you are correct, 0'' is valid, it returns 39

Valid or invalid?  Although I suppose the regex currently allows it either way, LOL... and I'm not even sure if that is bad or not.

File: src/languages/yaml.js
Comment (allejo) [reply src/languages/yaml.js:105]: Taking this snippet out of context, I got the following screenshots:

image
image
image

`\S` is short for [`[^ \f\n\r\t\v\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff]`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Character_Classes)

At the very least, shouldn't `\t` be inside alongside `\n` too?

File: src/languages/prolog.js
Comment (allejo) [reply src/languages/prolog.js:60]: My bad, I misread your original comment 😅 `0''` is _valid_ as seen in the screenshot. But since the regex allows it, works for me

File: src/languages/yaml.js
Comment (joshgoebel) [reply src/languages/yaml.js:105]: That's clear enough, but I think the prior behavior is WRONG... (unless you want to school me on a finer detail of YAML here - which is possible since YAML is insane)... the intention AFAIK is to go all the way to the end of the line... I'm not sure tabs should break a string sequence... I would expect them to simply be registered as part of the content.

If that is *incorrect* due to some insanity of YAML then I believe you would be right in that we should add `\t`.

Comment (allejo) [reply src/languages/yaml.js:105]: Ahhh I see what this regex is supposed to do now. I was just comparing the semantics of diffs between regexes. You are correct, there should not be a `\t` in here and previous behavior was wrong 👍

Comment (allejo) [inline src/languages/yaml.js:105]:
```suggestion
      begin: '[\\|>]([0-9]?[+-])?[ ]*\\n( +)[^\\n]+\\n(\\2[^\\n]+\\n?)*'
```

While we're talking about the behavior of this regex in this PR, should it be one or more spaces for the indentation? Having no indentation is invalid.

```yaml
key: |
multi-string
value
```

Comment (joshgoebel) [reply src/languages/yaml.js:105]: In this case (and what I know of YAML) I'd agree but I'm not certain what's going on with all the other `[\\|>]([0-9]?[+-])?` there... are there some other variants we're not considering?  I think the 0-9 is to allow offsets/extra indents but I'm not sure what `\\` and `>`... is the \` is just needless escaping then that leaves only the `>`... does that change anything?

Comment (allejo) [reply src/languages/yaml.js:105]: My favorite resource for deciphering YAML multi-line values... https://yaml-multiline.info/

- You can use 1-9 for indentation if the first line's indentation won't match the rest of the lines

According to yamllint.com, the `\` is valid,

this:

```yaml
key: \
  multi-string
  value
```

becomes this:

```yaml
key: "\\ multi-string value"
```

**Comment by joshgoebel** [comment]
Anything else? :)

**Review by allejo — APPROVED** [review]
Just don't forget to revert [this](https://github.com/highlightjs/highlight.js/pull/2759/files#r511647261) but this looks good to me

**Pull request merged**
